# Importing the Libraries


In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

# !conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


###### Data retrieved from City of Surrey Open Data 
https://data.surrey.ca/dataset/restaurants

In [2]:
df = pd.read_csv(r'D:\restaurants.csv')

In [3]:
df.head()

TRACKINGNUMBER             NAME PHYSICALADDRESS PHYSICALCITY     FACTYPE  \
0    SWOD-APSP3X  104 Sushi & Co.    10422 168 St       Surrey  Restaurant   
1    SDFO-88ESVK  5 Star Catering     5640 188 St       Surrey  Restaurant   
2    NDAA-8RNNVR    555 Pizza Ltd     9192 120 St       Surrey  Restaurant   
3    SYOG-5M5942  7-Eleven #26365     7986 120 St       Surrey  Restaurant   
4    GEDS-6KHW57  7-Eleven #26517    17638 60 Ave       Surrey  Restaurant   

    LATITUDE   LONGITUDE  
0  49.192059 -122.756256  
1  49.105449 -122.701315  
2  49.170274 -122.889988  
3  49.148189 -122.890221  
4  49.111322 -122.734271

In [6]:
pwd

'C:\\Users\\srira'

# Data Wrangling

In [4]:
df.drop('TRACKINGNUMBER',axis=1,inplace = True)

In [5]:
df.head()

NAME PHYSICALADDRESS PHYSICALCITY     FACTYPE   LATITUDE  \
0  104 Sushi & Co.    10422 168 St       Surrey  Restaurant  49.192059   
1  5 Star Catering     5640 188 St       Surrey  Restaurant  49.105449   
2    555 Pizza Ltd     9192 120 St       Surrey  Restaurant  49.170274   
3  7-Eleven #26365     7986 120 St       Surrey  Restaurant  49.148189   
4  7-Eleven #26517    17638 60 Ave       Surrey  Restaurant  49.111322   

    LONGITUDE  
0 -122.756256  
1 -122.701315  
2 -122.889988  
3 -122.890221  
4 -122.734271

In [6]:
df.rename(columns={'PHYSICALADDRESS':'ADDRESS','PHYSICALCITY':'CITY'},inplace = True)

In [7]:
df.head()

NAME       ADDRESS    CITY     FACTYPE   LATITUDE   LONGITUDE
0  104 Sushi & Co.  10422 168 St  Surrey  Restaurant  49.192059 -122.756256
1  5 Star Catering   5640 188 St  Surrey  Restaurant  49.105449 -122.701315
2    555 Pizza Ltd   9192 120 St  Surrey  Restaurant  49.170274 -122.889988
3  7-Eleven #26365   7986 120 St  Surrey  Restaurant  49.148189 -122.890221
4  7-Eleven #26517  17638 60 Ave  Surrey  Restaurant  49.111322 -122.734271

In [8]:
df.drop('FACTYPE',axis=1,inplace = True)

In [36]:
df = df.head(500)
df.shape

(500, 5)

In [37]:
address = 'Surrey, Canada'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of the City of Toronto are {}, {}.'.format(latitude, longitude))

C:\Users\srira\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: DeprecationWarning: Using Nominatim with the default "geopy/1.20.0" `user_agent` is strongly discouraged, as it violates Nominatim's ToS https://operations.osmfoundation.org/policies/nominatim/ and may possibly cause 403 and 429 HTTP errors. Please specify a custom `user_agent` with `Nominatim(user_agent="my-application")` or by overriding the default `user_agent`: `geopy.geocoders.options.default_user_agent = "my-application"`. In geopy 2.0 this will become an exception.
  This is separate from the ipykernel package so we can avoid doing imports until


The geograpical coordinate of the City of Toronto are 49.1913033, -122.8491439.


In [11]:
map_surrey = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, name in zip(df['LATITUDE'], df['LONGITUDE'], df['NAME']):
    label = '{}'.format(name)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=3,
        popup=label,
        color='green',
        fill=True,
        fill_color='#3199cc',
        fill_opacity=0.3,
        parse_html=False).add_to(map_surrey)  
    
map_surrey

##### Getting the foursquare credentials

In [1]:
CLIENT_ID = '' # your Foursquare ID
CLIENT_SECRET = '' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 
CLIENT_SECRET:


# Exploratory Data Analysis

##### Exploring the first restaurant and the nearby venues

In [39]:
# df.loc[0,'ADDRESS']
latitude = df.loc[1,'LATITUDE']
longitude = df.loc[1,'LONGITUDE']

##### Setting the foursquare request URL

In [40]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    latitude, 
    longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=ONAZS2PTG1GXEGVC5IFRCK4H3E5LLKODWWCXIBCQGI1VZBHC&client_secret=0XJQQBLBWNGYGWKLTNHOWDBWAAJMY53XOXOVF0WX3WRNCQ0R&v=20180605&ll=49.10544931,-122.70131471&radius=500&limit=100'

In [41]:
results = requests.get(url).json()
# results

In [42]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [43]:
venues = results['response']['groups'][0]['items']
    
nearby_venues1 = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues1 =nearby_venues1.loc[:, filtered_columns]

# filter the category for each row
nearby_venues1['venue.categories'] = nearby_venues1.apply(get_category_type, axis=1)

# clean columns
nearby_venues1.columns = [col.split(".")[-1] for col in nearby_venues1.columns]

nearby_venues1.head()

name   categories        lat         lng
0      The Clayton Public House          Pub  49.104655 -122.701060
1           Sunrise Golf Centre  Golf Course  49.105024 -122.701071
2  Mavericks Taphouse And Grill          Pub  49.104966 -122.701028

##### The above cell shows that there are three popular venues near the first restaurant ' 104 Sushi and Co' . There are two pubs and a golf course. Probably not a good idea to open a restaurant near noisy unpleasant pubs

##### Now lets get the nearby venues for all the restaurants from our Dataset

In [44]:
def getNearbyVenues(names, latitudes, longitudes, radius=150):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['RESTAURANT NAME', 
                  'Restaurant Latitude', 
                  'Restaurant Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [45]:
surrey_venues = getNearbyVenues(names=df['NAME'],
                                   latitudes=df['LATITUDE'],
                                   longitudes=df['LONGITUDE']
                                  )

104 Sushi & Co.
5 Star Catering
555 Pizza Ltd
7-Eleven #26365
7-Eleven #26517
7-Eleven #34339
7-Eleven #34501
7-Eleven #34633
7-Eleven #34907
7-Eleven #35770
7-Eleven #36022H
7-Eleven #37876H
7-Eleven #38206
7-Eleven (Canada Store #37206H)
7-Eleven (Canada Store #37865H)
7-Eleven (Canada Store #37866H)
7-Eleven Canada Store #36676
7-Eleven Store # 37868
7-ElevenStore and Gas Bar (Canada Store #37875H)
North Surrey Secondary School Cafeteria
Northview Golf & Country Club (Canal Kiosk)
Northview Golf & Country Club (Kitchen/Duffey's Sports Grill)
Northview Golf & Country Club (Palmer Dining Room)
Northview Golf & Country Club (Ridge Kiosk)
Notti Biscotti
Nutri-Dough Pizza
O My Tea Cafe
Occasions at the Pond
Ocean Park Donair
Ocean Park Pizza
Ocean Park Pizza & Village Pub
Ocean Park Pizza (72nd Ave)
Oh So Sweet Sugar Boutique
Ok Hee Bae Food Co
Okonomiyaki Pon
Old Surrey Restaurant
Omega Pizza & Wings
Omega Pizza & Wings (152nd St)
On Yuen Restaurant
One More Thai Restaurant
One Up Cafe


In [49]:
surrey_venues.head(10)

RESTAURANT NAME  Restaurant Latitude  Restaurant Longitude  \
0  104 Sushi & Co.            49.192059           -122.756256   
1  104 Sushi & Co.            49.192059           -122.756256   
2  104 Sushi & Co.            49.192059           -122.756256   
3  104 Sushi & Co.            49.192059           -122.756256   
4  5 Star Catering            49.105449           -122.701315   
5  5 Star Catering            49.105449           -122.701315   
6  5 Star Catering            49.105449           -122.701315   
7    555 Pizza Ltd            49.170274           -122.889988   
8  7-Eleven #26365            49.148189           -122.890221   
9  7-Eleven #26365            49.148189           -122.890221   

                          Venue  Venue Latitude  Venue Longitude  \
0                104 Sushi & Co       49.192245      -122.756318   
1                    Flying Cup       49.191845      -122.756420   
2             Fraser Glen Pizza       49.191401      -122.756039   
3                777 Bubble Tea       49.191919      -122.756211   
4      The Clayton Public House       49.104655      -122.701060   
5           Sunrise Golf Centre       49.105024      -122.701071   
6  Mavericks Taphouse And Grill       49.104966      -122.701028   
7                  Thai By Thai       49.169821      -122.889959   
8      The Keg Steakhouse + Bar       49.147487      -122.890022   
9              Cactus Club Cafe       49.146924      -122.890932   

        Venue Category  
0     Sushi Restaurant  
1          Coffee Shop  
2          Pizza Place  
3             Tea Room  
4                  Pub  
5          Golf Course  
6                  Pub  
7      Thai Restaurant  
8           Steakhouse  
9  American Restaurant

##### Here is where it gets interesting. For a restaurant, the nearby venues retrieved from Foursquare shoul return the same restaurant as a spot as well. Well, only if the restaurant is popular or  preferred. As you can see above,  for some restaurants like the '5 Star Catering', the nearby venues doesnt include them in the results. This could be because they might be just a 'take away' restaurant with no dine in facilities. In such cases, its a potential good spot without having to worry about competition.

In [47]:
surrey_venues.shape

(2613, 7)

#####  How many popular venues are around each restaurant?

In [48]:
surrey_venues.groupby('RESTAURANT NAME').count()

Restaurant Latitude  \
RESTAURANT NAME                                                           
104 Sushi & Co.                                                       4   
5 Star Catering                                                       3   
555 Pizza Ltd                                                         1   
7-Eleven #26365                                                       9   
7-Eleven #26517                                                       3   
7-Eleven #34339                                                       1   
7-Eleven #34501                                                       3   
7-Eleven #34633                                                      12   
7-Eleven #34907                                                       4   
7-Eleven #35770                                                       1   
7-Eleven #36022H                                                      4   
7-Eleven #37876H                                                      3   
7-Eleven #38206                                                       4   
7-Eleven (Canada Store #37206H)                                       4   
7-Eleven (Canada Store #37865H)                                       7   
7-Eleven (Canada Store #37866H)                                       4   
7-Eleven Canada Store #36676                                          1   
7-Eleven Store # 37868                                               12   
7-ElevenStore and Gas Bar (Canada Store #37875H)                      6   
777 Pizza & Donair                                                    4   
A & W #0755                                                           4   
A&W #0180                                                            12   
A&W #0315                                                             5   
A&W #0338                                                            11   
A&W #0339                                                             4   
A&W #0348                                                            14   
A&W #0389                                                             5   
A&W #0575                                                             5   
A&W #0735                                                             8   
A&W #0768                                                             8   
A&W #749                                                              8   
A&W Restaurant #0767                                                 10   
A&W Restaurant #0789                                                  2   
A&W Restaurant (Campbell Heights)                                     4   
A1 Coffee & Donuts                                                   12   
Aaron's Steak & Pizza                                                 3   
Academics preKindergarten                                             2   
Active Body Express                                                   4   
Adana Grill                                                           5   
Afghan Bakery and Fast Food                                           4   
Afghan Chopan                                                         2   
Afghan Kitchen                                                        4   
Aggarwal Sweets                                                       3   
Agoongyi Korean Restaurant                                            8   
Aji Sushi Family Kitchen                                              7   
Aka Sushi                                                             1   
Akasaka Japanese Restaurant (Surrey)                                  6   
Akash Chaat House & Restaurant                                        3   
Al & Jan's Fish & Chips                                               9   
Alebi African Cuisine                                                 3   
Alexandra Neighbourhood House (Main Dining Hall)                      4   
Aling Mary's Filipino Restaurant                                      4   
Al

In [50]:
print('There are {} uniques categories.'.format(len(surrey_venues['Venue Category'].unique())))

There are 163 uniques categories.


##### One hot Encoding the Venues

In [51]:
surrey_onehot = pd.get_dummies(surrey_venues[['Venue Category']], prefix="", prefix_sep="")

# # add neighborhood column back to dataframe
surrey_onehot['RESTAURANT NAME'] = surrey_venues['RESTAURANT NAME'] 

# # move neighborhood column to the first column
fixed_columns = [surrey_onehot.columns[-1]] + list(surrey_onehot.columns[:-1])
surrey_onehot = surrey_onehot[fixed_columns]

surrey_onehot.head(10)

RESTAURANT NAME  Afghan Restaurant  American Restaurant  \
0  104 Sushi & Co.                  0                    0   
1  104 Sushi & Co.                  0                    0   
2  104 Sushi & Co.                  0                    0   
3  104 Sushi & Co.                  0                    0   
4  5 Star Catering                  0                    0   
5  5 Star Catering                  0                    0   
6  5 Star Catering                  0                    0   
7    555 Pizza Ltd                  0                    0   
8  7-Eleven #26365                  0                    0   
9  7-Eleven #26365                  0                    1   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Dealership  \
0                    0                 0                   0                0   
1                    0                 0                   0                0   
2                    0                 0                   0                0   
3                    0                 0                   0                0   
4                    0                 0                   0                0   
5                    0                 0                   0                0   
6                    0                 0                   0                0   
7                    0                 0                   0                0   
8                    0                 0                   0                0   
9                    0                 0                   0                0   

   Automotive Shop  BBQ Joint  Bakery  Bank  Bar  Baseball Field  \
0                0          0       0     0    0               0   
1                0          0       0     0    0               0   
2                0          0       0     0    0               0   
3                0          0       0     0    0               0   
4                0          0       0     0    0               0   
5                0          0       0     0    0               0   
6                0          0       0     0    0               0   
7                0          0       0     0    0               0   
8                0          0       0     0    0               0   
9                0          0       0     0    0               0   

   Big Box Store  Bistro  Board Shop  Bookstore  Bowling Alley  \
0              0       0           0          0              0   
1              0       0           0          0              0   
2              0       0           0          0              0   
3              0       0           0          0              0   
4              0       0           0          0              0   
5              0       0           0          0              0   
6              0       0           0          0              0   
7              0       0           0          0              0   
8              0       0           0          0              0   
9              0       0           0          0              0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Burger Joint  \
0               0        0            0                0             0   
1               0        0            0                0             0   
2               0        0            0                0             0   
3               0        0            0                0             0   
4               0        0            0                0             0   
5               0        0            0                0             0   
6               0        0            0                0             0   
7               0        0            0                0             0   
8               0        0            0                0             0   
9               0        0            0                0             0   

   Bus Line  Bus Station  Bus Stop  Café  Campground  Carpet Store  \
0         0            0         0     0           0             0 

##### Grouping the results by the restaurants

In [52]:
surrey_grouped = surrey_onehot.groupby('RESTAURANT NAME').mean().reset_index()

In [53]:
surrey_grouped.head()

RESTAURANT NAME  Afghan Restaurant  American Restaurant  \
0  104 Sushi & Co.                0.0             0.000000   
1  5 Star Catering                0.0             0.000000   
2    555 Pizza Ltd                0.0             0.000000   
3  7-Eleven #26365                0.0             0.111111   
4  7-Eleven #26517                0.0             0.000000   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Dealership  \
0                  0.0               0.0                 0.0              0.0   
1                  0.0               0.0                 0.0              0.0   
2                  0.0               0.0                 0.0              0.0   
3                  0.0               0.0                 0.0              0.0   
4                  0.0               0.0                 0.0              0.0   

   Automotive Shop  BBQ Joint  Bakery  Bank  Bar  Baseball Field  \
0              0.0        0.0     0.0   0.0  0.0             0.0   
1              0.0        0.0     0.0   0.0  0.0             0.0   
2              0.0        0.0     0.0   0.0  0.0             0.0   
3              0.0        0.0     0.0   0.0  0.0             0.0   
4              0.0        0.0     0.0   0.0  0.0             0.0   

   Big Box Store  Bistro  Board Shop  Bookstore  Bowling Alley  \
0            0.0     0.0         0.0        0.0            0.0   
1            0.0     0.0         0.0        0.0            0.0   
2            0.0     0.0         0.0        0.0            0.0   
3            0.0     0.0         0.0        0.0            0.0   
4            0.0     0.0         0.0        0.0            0.0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Burger Joint  \
0             0.0      0.0          0.0              0.0           0.0   
1             0.0      0.0          0.0              0.0           0.0   
2             0.0      0.0          0.0              0.0           0.0   
3             0.0      0.0          0.0              0.0           0.0   
4             0.0      0.0          0.0              0.0           0.0   

   Bus Line  Bus Station  Bus Stop  Café  Campground  Carpet Store  \
0       0.0          0.0       0.0   0.0         0.0           0.0   
1       0.0          0.0       0.0   0.0         0.0           0.0   
2       0.0          0.0       0.0   0.0         0.0           0.0   
3       0.0          0.0       0.0   0.0         0.0           0.0   
4       0.0          0.0       0.0   0.0         0.0           0.0   

   Cheese Shop  Chinese Restaurant  Chocolate Shop  Church  Clothing Store  \
0          0.0                 0.0             0.0     0.0             0.0   
1          0.0                 0.0             0.0     0.0             0.0   
2          0.0                 0.0             0.0     0.0             0.0   
3          0.0                 0.0             0.0     0.0             0.0   
4          0.0                 0.0             0.0     0.0             0.0   

   Coffee Shop  College Rec Center  Construction & Landscaping  \
0     0.250000                 0.0                         0.0   
1     0.000000                 0.0                         0.0   
2     0.000000                 0.0                         0.0   
3     0.111111                 0.0                         0.0   
4     0.000000                 0.0                         0.0   

   Convenience Store  Cosmetics Shop  Costume Shop  Deli / Bodega  \
0           0.000000             0.0           0.0            0.0   
1           0.000000             0.0           0.0            0.0   
2           0.000000             0.0           0.0            0.0   
3           0.111111             0.0           0.0            0.0   
4           0.333333             0.0           0.0            0.0   

   Department Store  Design Studio  Dim Sum Restaurant  Diner  Discount Store  \
0               0.0            0.0                 0.0    0.0             0.0   
1               0.0         

In [54]:
surrey_grouped.shape

(461, 164)

In [55]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

##### Sorting the venues 

In [57]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['RESTAURANT NAME']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
surrey_venues_sorted = pd.DataFrame(columns=columns)
surrey_venues_sorted['RESTAURANT NAME'] = surrey_grouped['RESTAURANT NAME']

for ind in np.arange(surrey_grouped.shape[0]):
    surrey_venues_sorted.iloc[ind, 1:] = return_most_common_venues(surrey_grouped.iloc[ind, :], num_top_venues)

surrey_venues_sorted.head()

RESTAURANT NAME 1st Most Common Venue 2nd Most Common Venue  \
0  104 Sushi & Co.              Tea Room           Pizza Place   
1  5 Star Catering                   Pub           Golf Course   
2    555 Pizza Ltd       Thai Restaurant         Event Service   
3  7-Eleven #26365            Restaurant   American Restaurant   
4  7-Eleven #26517           Flea Market     Convenience Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0           Coffee Shop      Sushi Restaurant     Electronics Store   
1           Yoga Studio         Event Service           Fish Market   
2           Flea Market           Fish Market     Fish & Chips Shop   
3     Other Repair Shop     Convenience Store      Greek Restaurant   
4           Gas Station         Event Service           Fish Market   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
1     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
2   Filipino Restaurant  Fast Food Restaurant        Farmers Market   
3     Indian Restaurant       Doctor's Office           Coffee Shop   
4     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0        Farmers Market     Falafel Restaurant  
1        Farmers Market     Falafel Restaurant  
2    Falafel Restaurant            Fabric Shop  
3            Steakhouse            Yoga Studio  
4        Farmers Market     Falafel Restaurant

# Model Development

##### Using K Means Clustering to cluster an segment the community based on the restaurants and the popular spots returned by foursquare

In [60]:
kclusters = 5

surrey_grouped_clustering = surrey_grouped.drop('RESTAURANT NAME', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(surrey_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:30] 

array([0, 4, 4, 4, 4, 4, 0, 4, 3, 4, 0, 3, 3, 0, 3, 3, 4, 4, 3, 4, 3, 3,
       3, 3, 3, 4, 4, 3, 0, 4])

In [65]:
surrey_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)



# surrey_merged.head()



In [64]:
surrey_venues_sorted.head()

Cluster Labels  RESTAURANT NAME 1st Most Common Venue  \
0               0  104 Sushi & Co.              Tea Room   
1               4  5 Star Catering                   Pub   
2               4    555 Pizza Ltd       Thai Restaurant   
3               4  7-Eleven #26365            Restaurant   
4               4  7-Eleven #26517           Flea Market   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Pizza Place           Coffee Shop      Sushi Restaurant   
1           Golf Course           Yoga Studio         Event Service   
2         Event Service           Flea Market           Fish Market   
3   American Restaurant     Other Repair Shop     Convenience Store   
4     Convenience Store           Gas Station         Event Service   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0     Electronics Store     Fish & Chips Shop   Filipino Restaurant   
1           Fish Market     Fish & Chips Shop   Filipino Restaurant   
2     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
3      Greek Restaurant     Indian Restaurant       Doctor's Office   
4           Fish Market     Fish & Chips Shop   Filipino Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Fast Food Restaurant        Farmers Market     Falafel Restaurant  
1  Fast Food Restaurant        Farmers Market     Falafel Restaurant  
2        Farmers Market    Falafel Restaurant            Fabric Shop  
3           Coffee Shop            Steakhouse            Yoga Studio  
4  Fast Food Restaurant        Farmers Market     Falafel Restaurant

### Cluster 1

##### This cluster is a cluster of yoga studios, stores, pizza and sandwich places, juice bars but mostly inclues the fish market nearby. What an useful insight !

In [66]:
surrey_venues_sorted.loc[surrey_venues_sorted['Cluster Labels'] == 0, surrey_venues_sorted.columns[[1] + list(range(5, surrey_venues_sorted.shape[1]))]]

RESTAURANT NAME  4th Most Common Venue  \
0                                 104 Sushi & Co.       Sushi Restaurant   
6                                 7-Eleven #34501            Yoga Studio   
10                               7-Eleven #36022H         Sandwich Place   
13                7-Eleven (Canada Store #37206H)     Chinese Restaurant   
28                                      A&W #0735            Pizza Place   
39                    Afghan Bakery and Fast Food      Electronics Store   
53                                  Ambros Subway         Sandwich Place   
54                   An Pho Vietnamese Restaurant    Japanese Restaurant   
57               Aomori Sushi Japanese Restaurant          Grocery Store   
70                                     BBQ Nation  Portuguese Restaurant   
80   Barcelos Flame Grilled Chicken (King George)      Electronics Store   
93                     Bento Nouveau (RCSS #1521)            Flea Market   
96                       Bhaia Sweet & Restaurant      Indian Restaurant   
98          Bharat Parmar Sweet Shop & Restaurant  Portuguese Restaurant   
109             Blenz Coffee (Central City South)            Pizza Place   
115                                 Boiling Point     Chinese Restaurant   
116                       Bolivar Heights Gas Bar            Fish Market   
119                      Bon Ga Korean Restaurant            Yoga Studio   
123                            Booster Juice #232            Pizza Place   
130                              Boston Pizza #59         Sandwich Place   
131                              Boston Pizza #61            Yoga Studio   
132                              Boston Pizza #62       Greek Restaurant   
133                              Boston Pizza #73         Discount Store   
143                                 O My Tea Cafe       Sushi Restaurant   
144               OPA Souvlaki of Greece (Surrey)            Pizza Place   
154                           Omega Pizza & Wings      Electronics Store   
155                Omega Pizza & Wings (152nd St)     Chinese Restaurant   
159                                 Orange Julius            Pizza Place   
162                        Otter Gas Bar @ Surrey       Sushi Restaurant   
167                Pacific Billiards and Snookers    Japanese Restaurant   
189                           Parkland Fellowship          Event Service   
190                          Parkland Food Market            Fish Market   
205                                Pho 99 Express            Fish Market   
207                      Pho Guildford Restaurant     Chinese Restaurant   
212                                    Pho Newton         Discount Store   
219                                       Pikanik       Greek Restaurant   
226                                   Pizza 64 #2            Fish Market   
230                      Pizza Factory (Surrey 2)          Event Service   
232                              Pizza Hut #31104    Japanese Restaurant   
235                              Pizza Hut #31123      Indian Restaurant   
240                                     Pizza Inn       Sushi Restaurant   
247                       Planet India Restaurant          Grocery Store   
276                      Rama Sweets & Restaurant            Yoga Studio   
281         Real Canadian Superstore #1521 - Deli            Flea Market   
293                    Ricky's Country Restaurant            Flea Market   
299                           Rokku Asian Cuisine       Greek Restaurant   
310               Royal Oak Pizza and Steak House          Grocery Store   
333                                        Saison            Pizza Place   
340                   Sandcastle Bowl Bar & Grill       Greek Restaurant   
362                               Scott's Landing            Fabric Shop   
378           Short & Sweet Patisserie and Bakery            Yoga Studio   
384                                 Sockeye Sushi       Sushi Restaurant   
398   

In [67]:
surrey_venues_sorted.loc[surrey_venues_sorted['Cluster Labels'] == 1, surrey_venues_sorted.columns[[1] + list(range(5, surrey_venues_sorted.shape[1]))]]

RESTAURANT NAME 4th Most Common Venue  \
137        Northview Golf & Country Club (Canal Kiosk)           Fish Market   
138  Northview Golf & Country Club (Kitchen/Duffey'...           Fish Market   
139  Northview Golf & Country Club (Palmer Dining R...           Fish Market   
140        Northview Golf & Country Club (Ridge Kiosk)           Fish Market   
196                    Peace Portal Golf Course Lounge           Fish Market   
197                Peace Portal Golf Course Restaurant           Fish Market   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
137     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
138     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
139     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
140     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
196     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
197     Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
137        Farmers Market    Falafel Restaurant            Fabric Shop  
138        Farmers Market    Falafel Restaurant            Fabric Shop  
139        Farmers Market    Falafel Restaurant            Fabric Shop  
140        Farmers Market    Falafel Restaurant            Fabric Shop  
196        Farmers Market    Falafel Restaurant            Fabric Shop  
197        Farmers Market    Falafel Restaurant            Fabric Shop

In [68]:
surrey_venues_sorted.loc[surrey_venues_sorted['Cluster Labels'] == 2, surrey_venues_sorted.columns[[1] + list(range(5, surrey_venues_sorted.shape[1]))]]

RESTAURANT NAME 4th Most Common Venue  \
40                             Afghan Chopan           Flea Market   
42                           Aggarwal Sweets         Event Service   
47            Akash Chaat House & Restaurant           Flea Market   
58                          Apna Chaat House     Electronics Store   
81                  Barney's Pizza & Chicken         Event Service   
83                              Basant Sweet         Event Service   
117                   Bollywood Banquet Hall         Event Service   
136  North Surrey Secondary School Cafeteria         Event Service   
149              Ocean Park Pizza (72nd Ave)           Yoga Studio   
193           Payal Chaat House & Restaurant           Flea Market   
224                        Pizza 24 (Surrey)           Yoga Studio   
253                       Popular Fish House           Yoga Studio   
264                      Punjab Banquet Hall         Event Service   
265                       Punjab Sweet House           Yoga Studio   
278                            Rana Catering         Shopping Mall   
309          Royal Curry Restaurant & Sweets           Yoga Studio   
311                               Royal Paan         Event Service   
314                            Royal's Pizza         Event Service   
319             Sachdeva Sweets & Restaurant     Electronics Store   
330                      Sahota's Live Grill           Fish Market   
342               Sapphire 2 Bistro Banquets     Electronics Store   
346       Satya Asha Fusion Bar & Restaurant           Yoga Studio   
359                         Scott Hill Pizza  Fast Food Restaurant   
361                         Scott's 96 Pizza         Event Service   
450                            Subway #30456     Electronics Store   

       5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
40               Fish Market     Fish & Chips Shop   Filipino Restaurant   
42               Fish Market     Fish & Chips Shop   Filipino Restaurant   
47               Fish Market     Fish & Chips Shop   Filipino Restaurant   
58         Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
81               Fish Market     Fish & Chips Shop   Filipino Restaurant   
83               Fish Market     Fish & Chips Shop   Filipino Restaurant   
117              Fish Market     Fish & Chips Shop   Filipino Restaurant   
136              Fish Market     Fish & Chips Shop   Filipino Restaurant   
149        Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
193              Fish Market     Fish & Chips Shop   Filipino Restaurant   
224            Event Service           Fish Market     Fish & Chips Shop   
253            Event Service           Fish Market     Fish & Chips Shop   
264              Fish Market     Fish & Chips Shop   Filipino Restaurant   
265        Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
278  North Indian Restaurant            Taco Place   Filipino Restaurant   
309            Event Service     Fish & Chips Shop   Filipino Restaurant   
311              Fish Market     Fish & Chips Shop   Filipino Restaurant   
314              Fish Market     Fish & Chips Shop   Filipino Restaurant   
319        Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
330        Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
342        Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
346        Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   
359         Sushi Restaurant    Chinese Restaurant     Fish & Chips Shop   
361              Fish Market     Fish & Chips Shop   Filipino Restaurant   
450        Fish & Chips Shop   Filipino Restaurant  Fast Food Restaurant   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
40   Fast Food Restaurant        Farmers Market     Falafel Restaurant  
42   Fast Food Restaurant        Farmers Market     Falafel Restaurant  
47   Fast Fo

In [69]:
surrey_venues_sorted.loc[surrey_venues_sorted['Cluster Labels'] == 4, surrey_venues_sorted.columns[[1] + list(range(5, surrey_venues_sorted.shape[1]))]]

RESTAURANT NAME  \
1                                      5 Star Catering   
2                                        555 Pizza Ltd   
3                                      7-Eleven #26365   
4                                      7-Eleven #26517   
5                                      7-Eleven #34339   
7                                      7-Eleven #34633   
9                                      7-Eleven #35770   
16                        7-Eleven Canada Store #36676   
17                              7-Eleven Store # 37868   
19                                  777 Pizza & Donair   
25                                           A&W #0348   
26                                           A&W #0389   
29                                           A&W #0768   
30                                            A&W #749   
32                                A&W Restaurant #0789   
34                                  A1 Coffee & Donuts   
35                               Aaron's Steak & Pizza   
43                          Agoongyi Korean Restaurant   
44                            Aji Sushi Family Kitchen   
45                                           Aka Sushi   
46                Akasaka Japanese Restaurant (Surrey)   
49                               Alebi African Cuisine   
50    Alexandra Neighbourhood House (Main Dining Hall)   
51                    Aling Mary's Filipino Restaurant   
52                                       All About Pho   
56                                Anjali's Dosa Garden   
59                                    Arbourside Court   
60                                         Archies BBQ   
62                           Arirang Korean Restaurant   
64                                 Asian Noodle Island   
66     Awahsuk Aboriginal Head Start Association of BC   
68                                    B & B Restaurant   
69                                BBQ King Delight Inc   
71                             Ba Le Deli & Sandwiches   
72                                    Bacayon Chibugan   
73                              Backstretch Restaurant   
74                           Baik Mi Korean Restaurant   
75                                   Bakeshop and Cafe   
76                                   Bamboo Restaurant   
78                             Barcelos - Impact Plaza   
84                                       Baselines Pub   
85                      Baskin-Robbins Ice Cream #5321   
86                                   Bear Creek Market   
87                          Bear Creek Park Concession   
88                      Beast & Brine Local Provisions   
89                                         Belles Cafe   
90                             Bengal Grill Restaurant   
91                             Bennett Craft & Kitchen   
92                                       Bento Nouveau   
94                          Bento Nouveau (RCSS #1556)   
97                       Bhaia Sweet Shop & Restaurant   
99                          Bharat Sweets & Restaurant   
100                                   Big Chicken Town   
101                          Big Ridge Brewing Company   
103                      Big Star Sandwich Co (Surrey)   
107                                       Blenz Coffee   
113                  Blenz Guildford Recreation Centre   
118                                Bombay Banquet Hall   
120                      Bonny Castle Childrens Center   
121       Bonnycastle Montessori Day Care Food Service   
122                                           Boom Bee   
124                                 Booster Juice #278   
125                                 Booster Juice #565   
126                         Booster Juice (Fraser Hwy)   
127                             Booster Juice (Surrey)   
128                                  Boston Pizza #020   
129                      Boston Pizza #18 - Nordel Way   
134                   Bourbon Street Grill (Guildford)   
145                                  

##### We can find the perfect spot for the new restaurant based on the above clusters